### ISSUE TITLE
Forgetting a reads clause allows to prove false (not a soundness bug, but room for improvement)

### ISSUE THREAD

**Issue by samuelgruetter** [dafny-lang/dafny#316]
The following file passes verification (but fails on compilation to C# because of unimplemented methods).

```
datatype Result<T> =
  | Success(value: T)
  | Failure(error: string)

trait StringMap {
    ghost var content: map<string, string> 

    method Put(key: string, value: string) 
        modifies this
        ensures this.content == old(this.content)[key := value]

    function method Get(key: string): (r: Result<string>)
        //reads this // commented out on purpose
        ensures match r
            case Success(value) => key in content && content[key] == value
            case Failure(e) => key !in content

    static method Empty() returns (res: StringMap) ensures 
        fresh(res) && res.content == map[]
}

method Test()
{
    var m := StringMap.Empty();
    m.Put("testkey", "testvalue");
    var r := m.Get("testkey");
    assert false;
}
```

Note that in the method `Test()`, we can prove `false`.
This is not a soundness bug, because we're assuming an infeasible spec for `StringMap.Get`: It's not possible to implement this method without reading `content`.

Once we try to implement `StringMap`, we will notice that it's infeasible, and fix the specification of `Get`, and everything will be fine.

However, I still believe Dafny should be improved in this scenario in a way which allows the user to catch this error as early as possible. It's particularly important when `StringMap` is not a `trait`, but an extern class, which is never implemented in Dafny, but only in C#.

I would like Dafny to behave as follows on this example: It should emit a verification error which says that the postcondition of `Get` does not have sufficient read permissions to talk about `content`. Would this be possible?

**Comment by robin-aws** [comment]
Agreed there's room for improvement, if we can catch traits that are impossible to implement. However, I'd rather address the unsoundness that arises in the extern case directly - see #461 and #463.
